# 6.1 Sourcing Open Data: IBM Watson Marketing Customer Value Data

## Table of Contents
### 1. Importing libraries and dataset
### 2. Data Wrangling
### 3. Renaming columns
### 4. Data Understanding
### 5. Exporting Data

## Importing Libraries & Data set


In [70]:
# import libraries 
import pandas as pd
import numpy as np
import os

In [71]:
# create path
path = r'C:\Users\Dell\Documents\IBM Watson Marketing Customer Value Data'

In [72]:
# import  customer_value dataset
df_customer_value = pd.read_csv(os.path.join(path, '02. Data', 'Original Data', 'customer_value.csv'), index_col = False)

In [73]:
df_customer_value.head()

Customer       State  Customer Lifetime Value Response  Coverage Education  \
0  BU79786  Washington              2763.519279       No     Basic  Bachelor   
1  QZ44356     Arizona              6979.535903       No  Extended  Bachelor   
2  AI49188      Nevada             12887.431650       No   Premium  Bachelor   
3  WW63253  California              7645.861827       No     Basic  Bachelor   
4  HB64268  Washington              2813.692575       No     Basic  Bachelor   

  Effective To Date EmploymentStatus Gender  Income  ...  \
0           2/24/11         Employed      F   56274  ...   
1           1/31/11       Unemployed      F       0  ...   
2           2/19/11         Employed      F   48767  ...   
3           1/20/11       Unemployed      M       0  ...   
4        02/03/2011         Employed      M   43836  ...   

  Months Since Policy Inception Number of Open Complaints  Number of Policies  \
0                             5                         0                   1   
1                            42                         0                   8   
2                            38                         0                   2   
3                            65                         0                   7   
4                            44                         0                   1   

      Policy Type        Policy  Renew Offer Type  Sales Channel  \
0  Corporate Auto  Corporate L3            Offer1          Agent   
1   Personal Auto   Personal L3            Offer3          Agent   
2   Personal Auto   Personal L3            Offer1          Agent   
3  Corporate Auto  Corporate L2            Offer1    Call Center   
4   Personal Auto   Personal L1            Offer1          Agent   

  Total Claim Amount  Vehicle Class Vehicle Size  
0         384.811147   Two-Door Car      Medsize  
1        1131.464935  Four-Door Car      Medsize  
2         566.472247   Two-Door Car      Medsize  
3         529.881344            SUV      Medsize  
4         138.130879  Four-Door Car      Medsize  

[5 rows x 24 columns]

In [74]:
df_customer_value.shape

(9134, 24)

## 02. Data Wrangling

In [75]:
df_customer_value.columns

Index(['Customer', 'State', 'Customer Lifetime Value', 'Response', 'Coverage',
       'Education', 'Effective To Date', 'EmploymentStatus', 'Gender',
       'Income', 'Location Code', 'Marital Status', 'Monthly Premium Auto',
       'Months Since Last Claim', 'Months Since Policy Inception',
       'Number of Open Complaints', 'Number of Policies', 'Policy Type',
       'Policy', 'Renew Offer Type', 'Sales Channel', 'Total Claim Amount',
       'Vehicle Class', 'Vehicle Size'],
      dtype='object')

In [76]:
# Check for uniqueness
# Convert list of dictionaries to DataFrame
df = pd.DataFrame(df_customer_value)

In [77]:
# Check for unique values in each column
non_unique_columns = df.columns[df.nunique() == 1]

# Print non-unique column names
if len(non_unique_columns) > 0:
    print("The values in the following column(s) are not unique:", non_unique_columns.tolist())
else:
    print("All values in the dataset are unique.")

All values in the dataset are unique.


In [78]:
# check for mixed types
for col in df_customer_value.columns.tolist():
  weird = (df_customer_value[[col]].applymap(type) != df_customer_value[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer_value[weird]) > 0:
    print (col)

C:\Users\DELL\AppData\Local\Temp\ipykernel_24376\1560310002.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer_value[[col]].applymap(type) != df_customer_value[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\DELL\AppData\Local\Temp\ipykernel_24376\1560310002.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer_value[[col]].applymap(type) != df_customer_value[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\DELL\AppData\Local\Temp\ipykernel_24376\1560310002.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer_value[[col]].applymap(type) != df_customer_value[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\DELL\AppData\Local\Temp\ipykernel_24376\1560310002.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer_value[[col]].applymap(type) != df_customer_

In [79]:
# Droping unwanted columns
df_customer_value=df_customer_value.drop(columns=['Education'])

In [80]:
# Checking for missing values
df_customer_value.isnull().sum()

Customer                         0
State                            0
Customer Lifetime Value          0
Response                         0
Coverage                         0
Effective To Date                0
EmploymentStatus                 0
Gender                           0
Income                           0
Location Code                    0
Marital Status                   0
Monthly Premium Auto             0
Months Since Last Claim          0
Months Since Policy Inception    0
Number of Open Complaints        0
Number of Policies               0
Policy Type                      0
Policy                           0
Renew Offer Type                 0
Sales Channel                    0
Total Claim Amount               0
Vehicle Class                    0
Vehicle Size                     0
dtype: int64

### No missing values

In [81]:
# Finding duplicates
df_dups = df_customer_value[df_customer_value.duplicated()]

In [82]:
df_dups

Empty DataFrame
Columns: [Customer, State, Customer Lifetime Value, Response, Coverage, Effective To Date, EmploymentStatus, Gender, Income, Location Code, Marital Status, Monthly Premium Auto, Months Since Last Claim, Months Since Policy Inception, Number of Open Complaints, Number of Policies, Policy Type, Policy, Renew Offer Type, Sales Channel, Total Claim Amount, Vehicle Class, Vehicle Size]
Index: []

[0 rows x 23 columns]

### No duplicates

# 03. Renaming Columns

In [83]:
df_customer_value.rename(columns= {'Effective To Date' : 'Policy Expiry Date'},inplace= True)

In [84]:
df_customer_value.rename(columns= {'Customer': 'Customer Id'},inplace= True)

# 04. Data Understanding

In [85]:
df_customer_value.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9134 entries, 0 to 9133
Data columns (total 23 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Customer Id                    9134 non-null   object 
 1   State                          9134 non-null   object 
 2   Customer Lifetime Value        9134 non-null   float64
 3   Response                       9134 non-null   object 
 4   Coverage                       9134 non-null   object 
 5   Policy Expiry Date             9134 non-null   object 
 6   EmploymentStatus               9134 non-null   object 
 7   Gender                         9134 non-null   object 
 8   Income                         9134 non-null   int64  
 9   Location Code                  9134 non-null   object 
 10  Marital Status                 9134 non-null   object 
 11  Monthly Premium Auto           9134 non-null   int64  
 12  Months Since Last Claim        9134 non-null   i

In [86]:
df_customer_value.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count              9134.000000   9134.000000           9134.000000   
mean               8004.940475  37657.380009             93.219291   
std                6870.967608  30379.904734             34.407967   
min                1898.007675      0.000000             61.000000   
25%                3994.251794      0.000000             68.000000   
50%                5780.182197  33889.500000             83.000000   
75%                8962.167041  62320.000000            109.000000   
max               83325.381190  99981.000000            298.000000   

       Months Since Last Claim  Months Since Policy Inception  \
count              9134.000000                    9134.000000   
mean                 15.097000                      48.064594   
std                  10.073257                      27.905991   
min                   0.000000                       0.000000   
25%                   6.000000                      24.000000   
50%                  14.000000                      48.000000   
75%                  23.000000                      71.000000   
max                  35.000000                      99.000000   

       Number of Open Complaints  Number of Policies  Total Claim Amount  
count                9134.000000         9134.000000         9134.000000  
mean                    0.384388            2.966170          434.088794  
std                     0.910384            2.390182          290.500092  
min                     0.000000            1.000000            0.099007  
25%                     0.000000            1.000000          272.258244  
50%                     0.000000            2.000000          383.945434  
75%                     0.000000            4.000000          547.514839  
max                     5.000000            9.000000         2893.239678

#### We cannot remove or neglect the values '0' income since a large proportion of the data contains missing values, removing them may result in a significant loss of information and bias the analysis.

In [87]:
df_customer_value.shape

(9134, 23)

In [88]:
df_customer_value.head()

Customer Id       State  Customer Lifetime Value Response  Coverage  \
0     BU79786  Washington              2763.519279       No     Basic   
1     QZ44356     Arizona              6979.535903       No  Extended   
2     AI49188      Nevada             12887.431650       No   Premium   
3     WW63253  California              7645.861827       No     Basic   
4     HB64268  Washington              2813.692575       No     Basic   

  Policy Expiry Date EmploymentStatus Gender  Income Location Code  ...  \
0            2/24/11         Employed      F   56274      Suburban  ...   
1            1/31/11       Unemployed      F       0      Suburban  ...   
2            2/19/11         Employed      F   48767      Suburban  ...   
3            1/20/11       Unemployed      M       0      Suburban  ...   
4         02/03/2011         Employed      M   43836         Rural  ...   

  Months Since Policy Inception  Number of Open Complaints  \
0                             5                          0   
1                            42                          0   
2                            38                          0   
3                            65                          0   
4                            44                          0   

   Number of Policies     Policy Type        Policy  Renew Offer Type  \
0                   1  Corporate Auto  Corporate L3            Offer1   
1                   8   Personal Auto   Personal L3            Offer3   
2                   2   Personal Auto   Personal L3            Offer1   
3                   7  Corporate Auto  Corporate L2            Offer1   
4                   1   Personal Auto   Personal L1            Offer1   

  Sales Channel Total Claim Amount  Vehicle Class Vehicle Size  
0         Agent         384.811147   Two-Door Car      Medsize  
1         Agent        1131.464935  Four-Door Car      Medsize  
2         Agent         566.472247   Two-Door Car      Medsize  
3   Call Center         529.881344            SUV      Medsize  
4         Agent         138.130879  Four-Door Car      Medsize  

[5 rows x 23 columns]

In [89]:
# showing the value counts of renewal offers
df_customer_value['Renew Offer Type'].value_counts()

Renew Offer Type
Offer1    3752
Offer2    2926
Offer3    1432
Offer4    1024
Name: count, dtype: int64

In [90]:
# filtering the DataFrame to include only rows where 'Response' is 'Yes
response_yes_df = df_customer_value[df_customer_value['Response'] == 'Yes']

In [91]:
# showing the value counts of renewal offers where 'Response' is Yes
response_yes_df['Renew Offer Type'].value_counts()

Renew Offer Type
Offer2    684
Offer1    594
Offer3     30
Name: count, dtype: int64

#### 
Out of a total of 9134 records, only 1308 individuals accepted the new policy offer. Additionally, offer 4 was not accepted by anyoneThis particular offer may require further examination or adjustments to increase its appeal..

In [92]:
# Group by demographic or policy type segment
grouped = df.groupby(['Gender', 'Policy Type'])

In [93]:
grouped['Customer Lifetime Value'].describe()

count         mean          std          min  \
Gender Policy Type                                                     
F      Corporate Auto  1005.0  7942.296743  6527.703049  2004.350666   
       Personal Auto   3454.0  8132.580041  7107.349236  1898.683686   
       Special Auto     199.0  8251.427881  6394.589754  2004.350666   
M      Corporate Auto   963.0  7680.944727  6141.798341  2129.063055   
       Personal Auto   3334.0  7918.359992  6908.708911  1898.007675   
       Special Auto     179.0  8975.365788  7542.595671  2191.133237   

                               25%          50%           75%          max  
Gender Policy Type                                                          
F      Corporate Auto  3808.122147  5678.050167   9031.214859  46770.94705  
       Personal Auto   4048.895576  5730.633840   9031.214859  73225.95652  
       Special Auto    4373.264920  6503.397049   9417.200225  38496.94701  
M      Corporate Auto  3968.079029  5793.157728   8804.898950  61134.68307  
       Personal Auto   3963.396956  5814.145650   8851.377572  83325.38119  
       Special Auto    4290.592156  6236.018595  10871.288965  44795.46942

#### There's considerable variability in customer lifetime values, as indicated by the standard deviations across different policy types and genders.The "Personal Auto" policy type appears to be the most popular among both genders, as it has the highest count for both males and females.Among females, the highest mean customer lifetime value is associated with the "Special Auto" policy type, followed by "Personal Auto" and "Corporate Auto."For males, the mean customer lifetime value is highest for the "Special Auto" policy type as well, with "Personal Auto" following closely behind, and "Corporate Auto" having the lowest mean value.There's a notable difference in the mean customer lifetime values between genders. For instance, females generally exhibit higher mean values compared to males across all policy types.

In [94]:
# 05. Export Data File as Pickle format
# Export data as pickle format to Prepared Data
df_customer_value.to_pickle(os.path.join(path, '02. Data', 'Prepared Data', 'customer_value_new.pkl'))